gordon/metrics/ffwd.py

```diff
@@ -0,0 +1,200 @@
+        self.loop = loop or asyncio.get_event_loop()
+
+    async def send(self, metric):
+        """Transform metric to JSON bytestring and send to server.
+
+        Args:
+            metric (dict): Complete metric to send as JSON.
+        """
+        message = json.dumps(metric).encode('utf-8')
+        await self.loop.create_datagram_endpoint(
+            lambda: UDPClientProtocol(message),
+            remote_addr=(self.ip, self.port))
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class FfwdTimer:
+    """Timer which sends UDP messages to FFWD on completion.
+
+    Args:
+        metric (dict): Dict representation of the metric to send.
+        udp_client (UDPClient): A metric sending client.
+        time_unit (number): (optional) Scale time unit for use with
+            time.time(), for example: 1E9 to send nanoseconds.
+    """
+    def __init__(self, metric, udp_client, time_unit=None):
+        self.metric = metric
+        self.udp_client = udp_client
+        self.time_unit = time_unit or 1
+
+    async def __aenter__(self):
+        """Enter context manager to start timing."""
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        """Exit context manager to stop timing."""
+        await self.stop()
+
+    async def start(self):
+        """Start timer."""
+        self._start_time = time.perf_counter()
+
+    async def stop(self):
+        """Stop timer."""
+        time_elapsed = time.perf_counter() - self._start_time
+        self.metric['value'] = time_elapsed * self.time_unit
+        await self.udp_client.send(self.metric)
+
+
+@zope.interface.implementer(interfaces.IMetricRelay)
+class SimpleFfwdRelay:
+    """Metrics relay which sends to FFWD immediately.
+
+    The relay does no client-side aggregation and metrics are
+    emitted immediately. The relay uses a combination of the `key and
+    attributes fields <https://github.com/spotify/ffwd/tree/master/
+    modules/json>`_ to semantically identify metrics in ffwd.
+
+    Args:
+        config (dict): Configuration with optional keys described above.
+    """
+    def __init__(self, config):
+        self.key = config.get('key', 'gordon-service')
+        self.udp_client = UDPClient(
+            config.get('ffwd_ip'), config.get('ffwd_port'))
+        self.time_unit = config.get('time_unit')
+
+    def _create_metric(self, metric_name, value, context, **kwargs):
+        attrs = context or {}
+        attrs['what'] = metric_name
+
+        metric = {
+            'key': self.key,
+            'attributes': attrs,
+            'value': value,
+            'type': 'metric'
+        }
+        return metric
+
+    async def incr(self, metric_name, value=1, context=None, **kwargs):
+        """Increase a metric by 1 or a given amount.
+
+        Args:
+            metric_name (str): Identifier of the metric.
+            value (int): (optional) Value with which to increase the metric.
```

Comment:
Nitpick: "default to ``1``"?

Reply:
I thought saying optional was sufficient, since if nothing is provided, it isn't scaled (i.e. it just happens we use 1 for that), but I can add it.

Reply:
Oh sorry, still had the previous commented line in mind, when I answered. Still gonna add it, but I feel less conflicted about it now :)